internal/topology/upgrade/clusterctl_upgrade_test.go

```diff
@@ -0,0 +1,1174 @@
+				ClusterClassTemplate: clusterv1.ClusterClassTemplate{
+					Ref: &clusterv1.ClusterClassTemplateReference{
+						Kind:       "TestResourceTemplate",
+						Name:       controlPlaneTemplate.Name,
+						APIVersion: testt2v1beta2.GroupVersion.String(),
+					},
+				},
+				MachineInfrastructure: &clusterv1.ClusterClassTemplate{
+					Ref: &clusterv1.ClusterClassTemplateReference{
+						Kind:       "TestResourceTemplate",
+						Name:       infrastructureClusterTemplate1.Name,
+						APIVersion: testt2v1beta2.GroupVersion.String(),
+					},
+				},
+			},
+			Workers: clusterv1.WorkersClass{
+				MachineDeployments: []clusterv1.MachineDeploymentClass{
+					machineDeploymentClass1,
+				},
+			},
+			Patches: []clusterv1.ClusterClassPatch{
+				{
+					// Add the omitable fields.
+					// Note: the omitable fields are not dropped because the DefaulterRemoveUnknownOrOmitableFields options is not used anymore.
+					Name: "patch-t2",
+					Definitions: []clusterv1.PatchDefinition{
+						{
+							Selector: clusterv1.PatchSelector{
+								APIVersion: testt2v1beta2.GroupVersion.String(),
+								Kind:       "TestResourceTemplate",
+								MatchResources: clusterv1.PatchSelectorMatch{
+									InfrastructureCluster: true,
+									ControlPlane:          true,
+									MachineDeploymentClass: &clusterv1.PatchSelectorMatchMachineDeploymentClass{
+										Names: []string{machineDeploymentClass1.Class},
+									},
+								},
+							},
+							JSONPatches: []clusterv1.JSONPatch{
+								{
+									Op:    "add",
+									Path:  "/spec/template/spec/omitable",
+									Value: &apiextensionsv1.JSON{Raw: []byte(`""`)},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	g.Expect(env.CreateAndWait(ctx, clusterClass)).To(Succeed())
+	return clusterClass
+}
+
+// getClusterTopologyReferences gets all the references for a test cluster + the corresponding generation.
+// NOTE: it also stores all the objects / how they change across generation into allObjs for debugging.
+func getClusterTopologyReferences(cluster *clusterv1.Cluster, version string, additionalChecks func(*unstructured.Unstructured) error) (map[clusterv1.ContractVersionedObjectReference]int64, error) {
+	refs := map[clusterv1.ContractVersionedObjectReference]int64{}
+	// Get all the references
+
+	actualCluster := &clusterv1.Cluster{}
+	if err := env.Get(ctx, client.ObjectKeyFromObject(cluster), actualCluster); err != nil {
+		return nil, errors.Wrapf(err, "failed to get cluster %s", cluster.Name)
+	}
+	if c := conditions.Get(actualCluster, clusterv1.ClusterTopologyReconciledCondition); c != nil && c.Status != metav1.ConditionTrue {
+		return nil, errors.Errorf("cluster %s topology is not reconciled", cluster.Name)
+	}
+
+	addToAllObj(cluster.Name, "cluster", actualCluster)
+	refs[clusterv1.ContractVersionedObjectReference{
+		APIGroup: actualCluster.GroupVersionKind().Group,
+		Kind:     actualCluster.GroupVersionKind().Kind,
```

Comment:
Let's please use the group constant from the API package and hard-code kind.

It's a fragile assumption that apiVersion + Kind are set in a typed object (this doesn't work consistently across cached client / live client and fake client, there were a few PRs in CR to consistently not set them anymore on typed objects)